File: tests/relaunch/pumpAmm.ts
import {
  Keypair,
  PublicKey,
  SystemProgram,
  TransactionInstruction,
} from "@solana/web3.js";
import * as token from "@solana/spl-token";
import { BanksClient, ProgramTestContext } from "solana-bankrun";
import {
  getPumpCreatorVaultAuthorityAddr,
  getPumpPoolV2Addr,
  parsePumpGlobalConfig,
  PumpGlobalConfigAccount,
  PUMP_AMM_EVENT_AUTHORITY,
  PUMP_AMM_FEE_CONFIG,
  PUMP_AMM_GLOBAL_CONFIG,
  PUMP_AMM_PROGRAM_ID,
  PUMP_FEES_PROGRAM_ID,
  PUMP_PROGRAM_ID,
} from "@metadaoproject/programs";

const TOKEN_ACCOUNT_RENT = 2_039_280n;

export const PUMP_GLOBAL_VOLUME_ACCUMULATOR = PublicKey.findProgramAddressSync(
  [Buffer.from("global_volume_accumulator")],
  PUMP_AMM_PROGRAM_ID,
)[0];

// Stable coin creator used by fabricated pools so creator-vault ATAs are
// deterministic across tests.
export const PUMP_TEST_COIN_CREATOR = Keypair.fromSeed(
  new Uint8Array(Buffer.from("relaunch-fixture-coin-creator!!!")),
).publicKey;

export function getPumpPoolAuthorityAddr(baseMint: PublicKey): PublicKey {
  return PublicKey.findProgramAddressSync(
    [Buffer.from("pool-authority"), baseMint.toBuffer()],
    PUMP_PROGRAM_ID,
  )[0];
}

export function getPumpPoolAddr({
  index,
  creator,
  baseMint,
  quoteMint,
}: {
  index: number;
  creator: PublicKey;
  baseMint: PublicKey;
  quoteMint: PublicKey;
}): [PublicKey, number] {
  const indexBuf = Buffer.alloc(2);
  indexBuf.writeUInt16LE(index);
  return PublicKey.findProgramAddressSync(
    [
      Buffer.from("pool"),
      indexBuf,
      creator.toBuffer(),
      baseMint.toBuffer(),
      quoteMint.toBuffer(),
    ],
    PUMP_AMM_PROGRAM_ID,
  );
}

export function getCanonicalPumpPoolAddr(
  baseMint: PublicKey,
  quoteMint: PublicKey,
): [PublicKey, number] {
  return getPumpPoolAddr({
    index: 0,
    creator: getPumpPoolAuthorityAddr(baseMint),
    baseMint,
    quoteMint,
  });
}

export function getUserVolumeAccumulatorAddr(user: PublicKey): PublicKey {
  return PublicKey.findProgramAddressSync(
    [Buffer.from("user_volume_accumulator"), user.toBuffer()],
    PUMP_AMM_PROGRAM_ID,
  )[0];
}

async function fetchGlobalConfig(
  banksClient: BanksClient,
): Promise<PumpGlobalConfigAccount> {
  const globalConfig = await banksClient.getAccount(PUMP_AMM_GLOBAL_CONFIG);
  return parsePumpGlobalConfig(Buffer.from(globalConfig!.data));
}

function packTokenAccount({
  mint,
  owner,
  amount,
}: {
  mint: PublicKey;
  owner: PublicKey;
  amount: bigint;
}): Buffer {
  const isNative = mint.equals(token.NATIVE_MINT);
  const data = Buffer.alloc(token.ACCOUNT_SIZE);
  token.AccountLayout.encode(
    {
      mint,
      owner,
      amount,
      delegateOption: 0,
      delegate: PublicKey.default,
      state: token.AccountState.Initialized,
      isNativeOption: isNative ? 1 : 0,
      isNative: isNative ? TOKEN_ACCOUNT_RENT : 0n,
      delegatedAmount: 0n,
      closeAuthorityOption: 0,
      closeAuthority: PublicKey.default,
    },
    data,
  );
  return data;
}

// Writes a token account at `address` via setAccount. WSOL accounts get
// amount-backed lamports so the token program can move lamports alongside
// native transfers.
export function writeTokenAccount(
  context: ProgramTestContext,
  {
    address,
    mint,
    owner,
    amount,
    tokenProgram = token.TOKEN_PROGRAM_ID,
  }: {
    address: PublicKey;
    mint: PublicKey;
    owner: PublicKey;
    amount: bigint;
    tokenProgram?: PublicKey;
  },
) {
  const isNative = mint.equals(token.NATIVE_MINT);
  context.setAccount(address, {
    data: packTokenAccount({ mint, owner, amount }),
    owner: tokenProgram,
    lamports: Number(TOKEN_ACCOUNT_RENT + (isNative ? amount : 0n)),
    executable: false,
  });
}

export type WritePumpPoolParams = {
  context: ProgramTestContext;
  baseMint: PublicKey;
  quoteMint: PublicKey;
  baseReserve: bigint;
  quoteReserve: bigint;
  baseTokenProgram?: PublicKey;
  // Overrides below fabricate non-canonical pools for negative tests.
  index?: number;
  creator?: PublicKey;
  owner?: PublicKey;
  coinCreator?: PublicKey;
};

export type PumpPool = {
  pool: PublicKey;
  baseMint: PublicKey;
  quoteMint: PublicKey;
  baseTokenProgram: PublicKey;
  poolBaseTokenAccount: PublicKey;
  poolQuoteTokenAccount: PublicKey;
  coinCreator: PublicKey;
};

// Pool account layout
//
//   offset size field                     type
//        0    8 discriminator             [241,154,109,4,17,177,109,188]
//        8    1 pool_bump                 u8
//        9    2 index                     u16 LE (0 = canonical)
//       11   32 creator                   Pubkey (pool-authority PDA when canonical)
//       43   32 base_mint                 Pubkey
//       75   32 quote_mint                Pubkey
//      107   32 lp_mint                   Pubkey (["pool_lp_mint", pool] PDA)
//      139   32 pool_base_token_account   Pubkey (ATA of pool, base token program)
//      171   32 pool_quote_token_account  Pubkey (ATA of pool, classic SPL)
//      203    8 lp_supply                 u64 LE (unread by swaps)
//      211   32 coin_creator              Pubkey
//      243    1 is_mayhem_mode            bool
//      244    1 is_cashback_coin          bool
//      245   16 virtual_quote_reserves    i128 LE (0 = legacy behavior)
//      261   40 (reserved)                zero on mainnet pools
//      301      total
//
// Fabricates a pump_amm Pool plus funded vaults directly via setAccount.
// Deliberate even with the real program loaded: only the pump bonding-curve
// program can sign for the canonical creator PDA, so a real create_pool can
// never produce the canonical fingerprint in tests.
export async function writePumpPool({
  context,
  baseMint,
  quoteMint,
  baseReserve,
  quoteReserve,
  baseTokenProgram = token.TOKEN_PROGRAM_ID,
  index = 0,
  creator,
  owner = PUMP_AMM_PROGRAM_ID,
  coinCreator = PUMP_TEST_COIN_CREATOR,
}: WritePumpPoolParams): Promise<PumpPool> {
  creator = creator ?? getPumpPoolAuthorityAddr(baseMint);
  const [pool, poolBump] = getPumpPoolAddr({
    index,
    creator,
    baseMint,
    quoteMint,
  });

  const poolBaseTokenAccount = token.getAssociatedTokenAddressSync(
    baseMint,
    pool,
    true,
    baseTokenProgram,
  );
  const poolQuoteTokenAccount = token.getAssociatedTokenAddressSync(
    quoteMint,
    pool,
    true,
  );
  const lpMint = PublicKey.findProgramAddressSync(
    [Buffer.from("pool_lp_mint"), pool.toBuffer()],
    PUMP_AMM_PROGRAM_ID,
  )[0];

  const data = Buffer.alloc(301);
  let offset = 0;
  Buffer.from([241, 154, 109, 4, 17, 177, 109, 188]).copy(data, offset); // Pool discriminator
  offset += 8;
  data.writeUInt8(poolBump, offset);
  offset += 1;
  data.writeUInt16LE(index, offset);
  offset += 2;
  for (const key of [
    creator,
    baseMint,
    quoteMint,
    lpMint,
    poolBaseTokenAccount,
    poolQuoteTokenAccount,
  ]) {
    key.toBuffer().copy(data, offset);
    offset += 32;
  }
  data.writeBigUInt64LE(1_000_000_000n, offset); // lp_supply, unread by swaps
  offset += 8;
  coinCreator.toBuffer().copy(data, offset);
  offset += 32;
  // is_mayhem_mode, is_cashback_coin, virtual_quote_reserves and the 40
  // reserved tail bytes stay zeroed.

  context.setAccount(pool, {
    data,
    owner,
    lamports: 2_985_840,
    executable: false,
  });

  writeTokenAccount(context, {
    address: poolBaseTokenAccount,
    mint: baseMint,
    owner: pool,
    amount: baseReserve,
    tokenProgram: baseTokenProgram,
  });
  writeTokenAccount(context, {
    address: poolQuoteTokenAccount,
    mint: quoteMint,
    owner: pool,
    amount: quoteReserve,
  });

  // Fee destinations must exist before swaps: the coin creator's quote vault
  // ATA and the protocol fee recipient's quote ATA. Skip any that another
  // fabricated pool already wrote so accrued balances survive.
  const creatorVaultAta = token.getAssociatedTokenAddressSync(
    quoteMint,
    getPumpCreatorVaultAuthorityAddr(coinCreator),
    true,
  );
  if (!(await context.banksClient.getAccount(creatorVaultAta))) {
    writeTokenAccount(context, {
      address: creatorVaultAta,
      mint: quoteMint,
      owner: getPumpCreatorVaultAuthorityAddr(coinCreator),
      amount: 0n,
    });
  }
  const { protocolFeeRecipients, buybackFeeRecipients } =
    await fetchGlobalConfig(context.banksClient);
  for (const recipient of [protocolFeeRecipients[0], ...buybackFeeRecipients]) {
    const recipientAta = token.getAssociatedTokenAddressSync(
      quoteMint,
      recipient,
      true,
    );
    if (!(await context.banksClient.getAccount(recipientAta))) {
      writeTokenAccount(context, {
        address: recipientAta,
        mint: quoteMint,
        owner: recipient,
        amount: 0n,
      });
    }
  }

  return {
    pool,
    baseMint,
    quoteMint,
    baseTokenProgram,
    poolBaseTokenAccount,
    poolQuoteTokenAccount,
    coinCreator,
  };
}

function swapAccountMetas(
  pool: PumpPool,
  user: PublicKey,
  protocolFeeRecipient: PublicKey,
) {
  const userBaseTokenAccount = token.getAssociatedTokenAddressSync(
    pool.baseMint,
    user,
    true,
    pool.baseTokenProgram,
  );
  const userQuoteTokenAccount = token.getAssociatedTokenAddressSync(
    pool.quoteMint,
    user,
    true,
  );
  const protocolFeeRecipientTokenAccount = token.getAssociatedTokenAddressSync(
    pool.quoteMint,
    protocolFeeRecipient,
    true,
  );
  const coinCreatorVaultAuthority = getPumpCreatorVaultAuthorityAddr(
    pool.coinCreator,
  );
  const coinCreatorVaultAta = token.getAssociatedTokenAddressSync(
    pool.quoteMint,
    coinCreatorVaultAuthority,
    true,
  );

  return [
    { pubkey: pool.pool, isSigner: false, isWritable: true },
    { pubkey: user, isSigner: true, isWritable: true },
    { pubkey: PUMP_AMM_GLOBAL_CONFIG, isSigner: false, isWritable: false },
    { pubkey: pool.baseMint, isSigner: false, isWritable: false },
    { pubkey: pool.quoteMint, isSigner: false, isWritable: false },
    { pubkey: userBaseTokenAccount, isSigner: false, isWritable: true },
    { pubkey: userQuoteTokenAccount, isSigner: false, isWritable: true },
    { pubkey: pool.poolBaseTokenAccount, isSigner: false, isWritable: true },
    { pubkey: pool.poolQuoteTokenAccount, isSigner: false, isWritable: true },
    { pubkey: protocolFeeRecipient, isSigner: false, isWritable: false },
    {
      pubkey: protocolFeeRecipientTokenAccount,
      isSigner: false,
      isWritable: true,
    },
    { pubkey: pool.baseTokenProgram, isSigner: false, isWritable: false },
    { pubkey: token.TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
    { pubkey: SystemProgram.programId, isSigner: false, isWritable: false },
    {
      pubkey: token.ASSOCIATED_TOKEN_PROGRAM_ID,
      isSigner: false,
      isWritable: false,
    },
    { pubkey: PUMP_AMM_EVENT_AUTHORITY, isSigner: false, isWritable: false },
    { pubkey: PUMP_AMM_PROGRAM_ID, isSigner: false, isWritable: false },
    { pubkey: coinCreatorVaultAta, isSigner: false, isWritable: true },
    { pubkey: coinCreatorVaultAuthority, isSigner: false, isWritable: false },
  ];
}

// sell(base_amount_in, min_quote_amount_out) — 21 accounts plus the same
// remaining-account tail as buy: pool_v2, then a buyback fee recipient with
// its quote ATA (mirrors pump's own SDK; required once buyback fees have
// accrued).
export function pumpSellIx({
  pool,
  user,
  protocolFeeRecipient,
  buybackFeeRecipient,
  baseAmountIn,
  minQuoteAmountOut,
}: {
  pool: PumpPool;
  user: PublicKey;
  protocolFeeRecipient: PublicKey;
  buybackFeeRecipient: PublicKey;
  baseAmountIn: bigint;
  minQuoteAmountOut: bigint;
}): TransactionInstruction {
  const data = Buffer.alloc(8 + 8 + 8);
  Buffer.from([51, 230, 133, 164, 1, 127, 131, 173]).copy(data, 0);
  data.writeBigUInt64LE(baseAmountIn, 8);
  data.writeBigUInt64LE(minQuoteAmountOut, 16);

  const keys = [
    ...swapAccountMetas(pool, user, protocolFeeRecipient),
    { pubkey: PUMP_AMM_FEE_CONFIG, isSigner: false, isWritable: false },
    { pubkey: PUMP_FEES_PROGRAM_ID, isSigner: false, isWritable: false },
    {
      pubkey: getPumpPoolV2Addr(pool.baseMint),
      isSigner: false,
      isWritable: false,
    },
    { pubkey: buybackFeeRecipient, isSigner: false, isWritable: false },
    {
      pubkey: token.getAssociatedTokenAddressSync(
        pool.quoteMint,
        buybackFeeRecipient,
        true,
      ),
      isSigner: false,
      isWritable: true,
    },
  ];

  return new TransactionInstruction({
    programId: PUMP_AMM_PROGRAM_ID,
    keys,
    data,
  });
}

// buy(base_amount_out, max_quote_amount_in, track_volume) — sell's accounts
// plus the two volume accumulators, 23 total, then the remaining accounts:
// pool_v2 and (required for buys, unlike sells) one buyback fee recipient
// with its quote ATA. Any member of the global config's buyback list works.
export function pumpBuyIx({
  pool,
  user,
  protocolFeeRecipient,
  buybackFeeRecipient,
  baseAmountOut,
  maxQuoteAmountIn,
  trackVolume = false,
}: {
  pool: PumpPool;
  user: PublicKey;
  protocolFeeRecipient: PublicKey;
  buybackFeeRecipient: PublicKey;
  baseAmountOut: bigint;
  maxQuoteAmountIn: bigint;
  trackVolume?: boolean;
}): TransactionInstruction {
  const data = Buffer.alloc(8 + 8 + 8 + 1);
  Buffer.from([102, 6, 61, 18, 1, 218, 235, 234]).copy(data, 0);
  data.writeBigUInt64LE(baseAmountOut, 8);
  data.writeBigUInt64LE(maxQuoteAmountIn, 16);
  data.writeUInt8(trackVolume ? 1 : 0, 24);

  const keys = [
    ...swapAccountMetas(pool, user, protocolFeeRecipient),
    {
      pubkey: PUMP_GLOBAL_VOLUME_ACCUMULATOR,
      isSigner: false,
      isWritable: false,
    },
    {
      pubkey: getUserVolumeAccumulatorAddr(user),
      isSigner: false,
      isWritable: true,
    },
    { pubkey: PUMP_AMM_FEE_CONFIG, isSigner: false, isWritable: false },
    { pubkey: PUMP_FEES_PROGRAM_ID, isSigner: false, isWritable: false },
    {
      pubkey: getPumpPoolV2Addr(pool.baseMint),
      isSigner: false,
      isWritable: false,
    },
    { pubkey: buybackFeeRecipient, isSigner: false, isWritable: false },
    {
      pubkey: token.getAssociatedTokenAddressSync(
        pool.quoteMint,
        buybackFeeRecipient,
        true,
      ),
      isSigner: false,
      isWritable: true,
    },
  ];

  return new TransactionInstruction({
    programId: PUMP_AMM_PROGRAM_ID,
    keys,
    data,
  });
}

export function pumpInitUserVolumeAccumulatorIx({
  payer,
  user,
}: {
  payer: PublicKey;
  user: PublicKey;
}): TransactionInstruction {
  return new TransactionInstruction({
    programId: PUMP_AMM_PROGRAM_ID,
    keys: [
      { pubkey: payer, isSigner: true, isWritable: true },
      { pubkey: user, isSigner: false, isWritable: false },
      {
        pubkey: getUserVolumeAccumulatorAddr(user),
        isSigner: false,
        isWritable: true,
      },
      { pubkey: SystemProgram.programId, isSigner: false, isWritable: false },
      { pubkey: PUMP_AMM_EVENT_AUTHORITY, isSigner: false, isWritable: false },
      { pubkey: PUMP_AMM_PROGRAM_ID, isSigner: false, isWritable: false },
    ],
    data: Buffer.from([94, 6, 202, 115, 255, 96, 232, 183]),
  });
}

File: programs/relaunch/src/pump_amm.rs
//! Types for reading pump_amm accounts, plus hand-built CPI builders —
//! pump_amm ships no Anchor-0.29 crate. Discriminators and args come from
//! pump_amm's IDL; the deployed program is newer than that IDL and requires a
//! remaining-accounts tail past the published list — the `pool_v2` PDA plus
//! one buyback fee recipient with its quote ATA (verified against live
//! mainnet swaps 2026-08-04).
use anchor_lang::prelude::*;
use anchor_lang::solana_program::instruction::{AccountMeta, Instruction};
use anchor_lang::solana_program::program::{invoke, invoke_signed};

use crate::error::RelaunchError;
use crate::pump_amm_program;

pub const POOL_DISCRIMINATOR: [u8; 8] = [241, 154, 109, 4, 17, 177, 109, 188];

pub const SELL_DISCRIMINATOR: [u8; 8] = [51, 230, 133, 164, 1, 127, 131, 173];

pub const BUY_DISCRIMINATOR: [u8; 8] = [102, 6, 61, 18, 1, 218, 235, 234];

pub const INIT_USER_VOLUME_ACCUMULATOR_DISCRIMINATOR: [u8; 8] =
    [94, 6, 202, 115, 255, 96, 232, 183];

/// The prefix of pump_amm's `Pool` account that canonicality validation
/// reads; trailing fields are ignored.
#[derive(AnchorDeserialize)]
pub struct PumpSwapPool {
    pub pool_bump: u8,
    pub index: u16,
    pub creator: Pubkey,
    pub base_mint: Pubkey,
    pub quote_mint: Pubkey,
}

impl PumpSwapPool {
    pub fn try_parse(data: &[u8]) -> Result<Self> {
        require!(
            data.len() > POOL_DISCRIMINATOR.len() && data[..8] == POOL_DISCRIMINATOR,
            RelaunchError::SourcePoolNotCanonical
        );
        Self::deserialize(&mut &data[8..])
            .map_err(|_| error!(RelaunchError::SourcePoolNotCanonical))
    }
}

/// The accounts of pump_amm's `sell`, in instruction order.
pub struct Sell<'info> {
    pub pool: AccountInfo<'info>,
    pub user: AccountInfo<'info>,
    pub global_config: AccountInfo<'info>,
    pub base_mint: AccountInfo<'info>,
    pub quote_mint: AccountInfo<'info>,
    pub user_base_token_account: AccountInfo<'info>,
    pub user_quote_token_account: AccountInfo<'info>,
    pub pool_base_token_account: AccountInfo<'info>,
    pub pool_quote_token_account: AccountInfo<'info>,
    pub protocol_fee_recipient: AccountInfo<'info>,
    pub protocol_fee_recipient_token_account: AccountInfo<'info>,
    pub base_token_program: AccountInfo<'info>,
    pub quote_token_program: AccountInfo<'info>,
    pub system_program: AccountInfo<'info>,
    pub associated_token_program: AccountInfo<'info>,
    pub event_authority: AccountInfo<'info>,
    pub program: AccountInfo<'info>,
    pub coin_creator_vault_ata: AccountInfo<'info>,
    pub coin_creator_vault_authority: AccountInfo<'info>,
    pub fee_config: AccountInfo<'info>,
    pub fee_program: AccountInfo<'info>,
    // remaining-accounts tail
    pub pool_v2: AccountInfo<'info>,
    pub buyback_fee_recipient: AccountInfo<'info>,
    pub buyback_fee_recipient_token_account: AccountInfo<'info>,
}

pub fn sell(
    accounts: Sell,
    base_amount_in: u64,
    min_quote_amount_out: u64,
    signer_seeds: &[&[&[u8]]],
) -> Result<()> {
    let mut data = Vec::with_capacity(24);
    data.extend_from_slice(&SELL_DISCRIMINATOR);
    data.extend_from_slice(&base_amount_in.to_le_bytes());
    data.extend_from_slice(&min_quote_amount_out.to_le_bytes());

    let metas = vec![
        AccountMeta::new(accounts.pool.key(), false),
        AccountMeta::new(accounts.user.key(), true),
        AccountMeta::new_readonly(accounts.global_config.key(), false),
        AccountMeta::new_readonly(accounts.base_mint.key(), false),
        AccountMeta::new_readonly(accounts.quote_mint.key(), false),
        AccountMeta::new(accounts.user_base_token_account.key(), false),
        AccountMeta::new(accounts.user_quote_token_account.key(), false),
        AccountMeta::new(accounts.pool_base_token_account.key(), false),
        AccountMeta::new(accounts.pool_quote_token_account.key(), false),
        AccountMeta::new_readonly(accounts.protocol_fee_recipient.key(), false),
        AccountMeta::new(accounts.protocol_fee_recipient_token_account.key(), false),
        AccountMeta::new_readonly(accounts.base_token_program.key(), false),
        AccountMeta::new_readonly(accounts.quote_token_program.key(), false),
        AccountMeta::new_readonly(accounts.system_program.key(), false),
        AccountMeta::new_readonly(accounts.associated_token_program.key(), false),
        AccountMeta::new_readonly(accounts.event_authority.key(), false),
        AccountMeta::new_readonly(accounts.program.key(), false),
        AccountMeta::new(accounts.coin_creator_vault_ata.key(), false),
        AccountMeta::new_readonly(accounts.coin_creator_vault_authority.key(), false),
        AccountMeta::new_readonly(accounts.fee_config.key(), false),
        AccountMeta::new_readonly(accounts.fee_program.key(), false),
        AccountMeta::new_readonly(accounts.pool_v2.key(), false),
        AccountMeta::new_readonly(accounts.buyback_fee_recipient.key(), false),
        AccountMeta::new(accounts.buyback_fee_recipient_token_account.key(), false),
    ];

    let account_infos = [
        accounts.pool,
        accounts.user,
        accounts.global_config,
        accounts.base_mint,
        accounts.quote_mint,
        accounts.user_base_token_account,
        accounts.user_quote_token_account,
        accounts.pool_base_token_account,
        accounts.pool_quote_token_account,
        accounts.protocol_fee_recipient,
        accounts.protocol_fee_recipient_token_account,
        accounts.base_token_program,
        accounts.quote_token_program,
        accounts.system_program,
        accounts.associated_token_program,
        accounts.event_authority,
        accounts.program,
        accounts.coin_creator_vault_ata,
        accounts.coin_creator_vault_authority,
        accounts.fee_config,
        accounts.fee_program,
        accounts.pool_v2,
        accounts.buyback_fee_recipient,
        accounts.buyback_fee_recipient_token_account,
    ];

    invoke_signed(
        &Instruction {
            program_id: pump_amm_program::id(),
            accounts: metas,
            data,
        },
        &account_infos,
        signer_seeds,
    )?;

    Ok(())
}

/// The accounts of pump_amm's `buy`, in instruction order.
pub struct Buy<'info> {
    pub pool: AccountInfo<'info>,
    pub user: AccountInfo<'info>,
    pub global_config: AccountInfo<'info>,
    pub base_mint: AccountInfo<'info>,
    pub quote_mint: AccountInfo<'info>,
    pub user_base_token_account: AccountInfo<'info>,
    pub user_quote_token_account: AccountInfo<'info>,
    pub pool_base_token_account: AccountInfo<'info>,
    pub pool_quote_token_account: AccountInfo<'info>,
    pub protocol_fee_recipient: AccountInfo<'info>,
    pub protocol_fee_recipient_token_account: AccountInfo<'info>,
    pub base_token_program: AccountInfo<'info>,
    pub quote_token_program: AccountInfo<'info>,
    pub system_program: AccountInfo<'info>,
    pub associated_token_program: AccountInfo<'info>,
    pub event_authority: AccountInfo<'info>,
    pub program: AccountInfo<'info>,
    pub coin_creator_vault_ata: AccountInfo<'info>,
    pub coin_creator_vault_authority: AccountInfo<'info>,
    pub global_volume_accumulator: AccountInfo<'info>,
    pub user_volume_accumulator: AccountInfo<'info>,
    pub fee_config: AccountInfo<'info>,
    pub fee_program: AccountInfo<'info>,
    // remaining-accounts tail
    pub pool_v2: AccountInfo<'info>,
    pub buyback_fee_recipient: AccountInfo<'info>,
    pub buyback_fee_recipient_token_account: AccountInfo<'info>,
}

pub fn buy(
    accounts: Buy,
    base_amount_out: u64,
    max_quote_amount_in: u64,
    track_volume: bool,
    signer_seeds: &[&[&[u8]]],
) -> Result<()> {
    let mut data = Vec::with_capacity(25);
    data.extend_from_slice(&BUY_DISCRIMINATOR);
    data.extend_from_slice(&base_amount_out.to_le_bytes());
    data.extend_from_slice(&max_quote_amount_in.to_le_bytes());
    data.push(track_volume as u8);

    let metas = vec![
        AccountMeta::new(accounts.pool.key(), false),
        AccountMeta::new(accounts.user.key(), true),
        AccountMeta::new_readonly(accounts.global_config.key(), false),
        AccountMeta::new_readonly(accounts.base_mint.key(), false),
        AccountMeta::new_readonly(accounts.quote_mint.key(), false),
        AccountMeta::new(accounts.user_base_token_account.key(), false),
        AccountMeta::new(accounts.user_quote_token_account.key(), false),
        AccountMeta::new(accounts.pool_base_token_account.key(), false),
        AccountMeta::new(accounts.pool_quote_token_account.key(), false),
        AccountMeta::new_readonly(accounts.protocol_fee_recipient.key(), false),
        AccountMeta::new(accounts.protocol_fee_recipient_token_account.key(), false),
        AccountMeta::new_readonly(accounts.base_token_program.key(), false),
        AccountMeta::new_readonly(accounts.quote_token_program.key(), false),
        AccountMeta::new_readonly(accounts.system_program.key(), false),
        AccountMeta::new_readonly(accounts.associated_token_program.key(), false),
        AccountMeta::new_readonly(accounts.event_authority.key(), false),
        AccountMeta::new_readonly(accounts.program.key(), false),
        AccountMeta::new(accounts.coin_creator_vault_ata.key(), false),
        AccountMeta::new_readonly(accounts.coin_creator_vault_authority.key(), false),
        AccountMeta::new_readonly(accounts.global_volume_accumulator.key(), false),
        AccountMeta::new(accounts.user_volume_accumulator.key(), false),
        AccountMeta::new_readonly(accounts.fee_config.key(), false),
        AccountMeta::new_readonly(accounts.fee_program.key(), false),
        AccountMeta::new_readonly(accounts.pool_v2.key(), false),
        AccountMeta::new_readonly(accounts.buyback_fee_recipient.key(), false),
        AccountMeta::new(accounts.buyback_fee_recipient_token_account.key(), false),
    ];

    let account_infos = [
        accounts.pool,
        accounts.user,
        accounts.global_config,
        accounts.base_mint,
        accounts.quote_mint,
        accounts.user_base_token_account,
        accounts.user_quote_token_account,
        accounts.pool_base_token_account,
        accounts.pool_quote_token_account,
        accounts.protocol_fee_recipient,
        accounts.protocol_fee_recipient_token_account,
        accounts.base_token_program,
        accounts.quote_token_program,
        accounts.system_program,
        accounts.associated_token_program,
        accounts.event_authority,
        accounts.program,
        accounts.coin_creator_vault_ata,
        accounts.coin_creator_vault_authority,
        accounts.global_volume_accumulator,
        accounts.user_volume_accumulator,
        accounts.fee_config,
        accounts.fee_program,
        accounts.pool_v2,
        accounts.buyback_fee_recipient,
        accounts.buyback_fee_recipient_token_account,
    ];

    invoke_signed(
        &Instruction {
            program_id: pump_amm_program::id(),
            accounts: metas,
            data,
        },
        &account_infos,
        signer_seeds,
    )?;

    Ok(())
}

/// The accounts of pump_amm's `init_user_volume_accumulator`, in instruction
/// order. Buys require the user's volume accumulator PDA to exist; this
/// creates it (rent paid by `payer`).
pub struct InitUserVolumeAccumulator<'info> {
    pub payer: AccountInfo<'info>,
    pub user: AccountInfo<'info>,
    pub user_volume_accumulator: AccountInfo<'info>,
    pub system_program: AccountInfo<'info>,
    pub event_authority: AccountInfo<'info>,
    pub program: AccountInfo<'info>,
}

pub fn init_user_volume_accumulator(accounts: InitUserVolumeAccumulator) -> Result<()> {
    let metas = vec![
        AccountMeta::new(accounts.payer.key(), true),
        AccountMeta::new_readonly(accounts.user.key(), false),
        AccountMeta::new(accounts.user_volume_accumulator.key(), false),
        AccountMeta::new_readonly(accounts.system_program.key(), false),
        AccountMeta::new_readonly(accounts.event_authority.key(), false),
        AccountMeta::new_readonly(accounts.program.key(), false),
    ];

    let account_infos = [
        accounts.payer,
        accounts.user,
        accounts.user_volume_accumulator,
        accounts.system_program,
        accounts.event_authority,
        accounts.program,
    ];

    invoke(
        &Instruction {
            program_id: pump_amm_program::id(),
            accounts: metas,
            data: INIT_USER_VOLUME_ACCUMULATOR_DISCRIMINATOR.to_vec(),
        },
        &account_infos,
    )?;

    Ok(())
}
